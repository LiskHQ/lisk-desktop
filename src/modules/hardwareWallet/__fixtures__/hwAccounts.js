export const hwAccounts = [
  {
    hw: {
      deviceId: '20231',
      model: 'Nano S',
      brand: 'Ledger',
    },
    metadata: {
      name: 'masoud123',
      pubkey: 'a07e06d3d21bfcbd6e6cb158a0adc18b6ccc92db66dc8a323c7462af9c539fab',
      path: '',
      accountIndex: 1,
      isHW: true,
      address: 'lskdgtenb76rf93bzd56cqn6ova46wfvoesbk4hnd',
      creationTime: '2022-12-08T15:30:39.979Z',
    },
    version: 1,
  },
  {
    hw: {
      deviceId: '20231',
      model: 'Nano S',
      brand: 'Ledger',
    },
    metadata: {
      name: 'masoud123',
      pubkey: 'a07e06d3d21bfcbd6e6cb158a0adc18b6ccc92db66dc8a323c7462af9c539fab',
      path: '',
      accountIndex: 2,
      isHW: true,
      address: 'lskdgtenb76rf93bzd56cqn6ova46wfvoesbk4hnd',
      creationTime: '2022-12-08T15:30:39.979Z',
    },
    version: 1,
  },
];
