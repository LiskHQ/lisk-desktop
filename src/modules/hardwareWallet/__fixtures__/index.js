export * from './mockHWDevices';
export * from './mockHWAccounts';
export * from './mockHWCurrentDevice';
