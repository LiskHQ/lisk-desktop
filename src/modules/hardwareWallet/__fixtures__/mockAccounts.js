export const mockAccounts = [
  {
    id: 1,
  },
  {
    id: 2,
  },
];
