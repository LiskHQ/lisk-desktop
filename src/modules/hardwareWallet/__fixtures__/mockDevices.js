export const mockDevices = [
  {
    id: 1,
  },
  {
    id: 2,
  },
];
