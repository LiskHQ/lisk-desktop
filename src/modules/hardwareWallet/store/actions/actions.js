import actionTypes from './actionTypes';

export const storeHWAccounts = (accounts) => ({
  type: actionTypes.storeHWAccounts,
  accounts,
});

export const removeHWAccounts = () => ({
  type: actionTypes.removeHWAccounts,
});
