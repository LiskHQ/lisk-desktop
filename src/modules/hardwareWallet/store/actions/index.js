export * from './accountsActions';
export * from './devicesActions';
