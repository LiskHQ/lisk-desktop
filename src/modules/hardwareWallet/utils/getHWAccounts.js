/* eslint-disable no-await-in-loop, max-statements */
import HWManager from '@hardwareWallet/manager/HWManager';
import { cryptography } from '@liskhq/lisk-client';
import { getCheckInitializedAccount } from '@account/utils/getCheckInitializedAccount';

export const getHWAccounts = async ({
  getName,
  device // fetch device from HWManager
}) => {
  const accounts = [];
  let accountIndex = 0;
  while (true) {
    const pubkey = await HWManager.getPublicKey(accountIndex);
    const address = cryptography.address.getAddressFromPublicKey(Buffer.from(pubkey, 'hex'));
    const config = { params: { address } };
    const isInitialized = await getCheckInitializedAccount({ config });
    if (!isInitialized) {
      accounts.push({
        hw: device,
        metadata: {
          address,
          pubkey,
          accountIndex,
          name: 'New account',
          path: '',
          isHW: true,
          creationTime: new Date().toISOString(),
        },
      });
      break;
    }
    accounts.push({
      hw: device,
      metadata: {
        address,
        pubkey,
        accountIndex,
        name: getName(address, device.model),
        path: '',
        isHW: true,
        creationTime: new Date().toISOString(),
      },
    });
    ++accountIndex;
  }
  return accounts;
};
