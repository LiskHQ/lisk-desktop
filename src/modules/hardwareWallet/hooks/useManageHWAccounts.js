import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { cryptography } from '@liskhq/lisk-client';
import HWManager from '@hardwareWallet/manager/HWManager';
import { selectSettings } from 'src/redux/selectors';
import { storeAccounts, removeAccounts } from '@hardwareWallet/store/actions/actions';
import useCheckInitializedAccount from '@common/hooks/useCheckInitializedAccount';

const getNameFromAccount = (address, settings) => {
  const { hardwareAccounts } = settings;
  const { deviceId } = HWManager.getActiveDeviceInfo();
  if (Array.isArray(hardwareAccounts[deviceId])) {
    const storedAccount = hardwareAccounts[deviceId].filter(
      (account) => account.address === address
    );
    return storedAccount.length ? storedAccount[0].name : null;
  }
  return null;
};

const useManageHWAccounts = () => {
  const dispatch = useDispatch();
  const settings = useSelector(selectSettings);
  const deviceInfo = HWManager.getActiveDeviceInfo();

  async function getAccounts() {
    const accounts = [];
    let accountIndex = 0;
    while (true) {
      // eslint-disable-next-line no-await-in-loop
      const pubkey = await HWManager.getPublicKey(accountIndex);
      // try {
      const address = cryptography.address.getAddressFromPublicKey(Buffer.from(pubkey, 'hex'));
      // eslint-disable-next-line no-await-in-loop
      const isInitialized = await useCheckInitializedAccount(address);
      if (!isInitialized) break;
      accounts.push({
        hw: deviceInfo,
        metadata: {
          address,
          pubkey,
          accountIndex,
          name: getNameFromAccount(address, settings),
          path: '',
          isHW: true,
          creationTime: new Date().toISOString(),
        },
      });
      ++accountIndex;
    }
    return accounts;
  }
  getAccounts().then((accList) => {
    dispatch(storeAccounts(accList));
  });
  useEffect(() => {
    if (deviceInfo.currentDeviceStatus === 'disconnected') {
      dispatch(removeAccounts());
    }
  }, [deviceInfo.currentDeviceStatus]);
};

export default useManageHWAccounts;
