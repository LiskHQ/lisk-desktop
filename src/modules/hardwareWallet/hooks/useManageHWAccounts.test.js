import { renderHook } from '@testing-library/react-hooks';
import { useSelector } from 'react-redux';
import { setHWAccounts, removeHWAccounts } from '../store/actions';
import { hwAccounts } from '../__fixtures__/hwAccounts';
import { getHWAccounts } from '../utils/getHWAccounts';
import useManageHWAccounts from './useManageHWAccounts';

jest.useRealTimers();

const mockDispatch = jest.fn();
const mockAppState = {
  hardwareWallet: {
    currentDevice: {
      deviceId: 20231,
      status: 'connected',
    },
  },
  settings: {
    hardwareAccounts: {},
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
  useSelector: jest.fn(),
}));

jest.mock('../utils/getHWAccounts');
getHWAccounts.mockResolvedValue(hwAccounts);

describe('useManageHWAccounts hook', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('stores the list of accounts in the hardware wallet', async () => {
    useSelector.mockImplementation((callback) => callback(mockAppState));
    const { waitFor } = renderHook(() => useManageHWAccounts());

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledTimes(1);
      const hwWalletAccountsDetails = setHWAccounts(hwAccounts);
      expect(mockDispatch).toHaveBeenCalledWith(expect.objectContaining(hwWalletAccountsDetails));
    });
  });

  it('removes the list of accounts when the device is disconnected', async () => {
    const updatedMockAppState = {
      hardwareWallet: {
        currentDevice: {
          deviceId: 20231,
          status: 'disconnected',
        },
        accounts: hwAccounts,
      },
      settings: {
        hardwareAccounts: {
          20231: [
            {
              address: 'a07e06d3d21bfcbd6e6cb158a0adc18b6ccc92db66dc8a323c7462af9c539fab',
              name: 'masoud123',
            },
            {
              address: '49dc55a205bf1fecbee41bde6e354eaa85fdc3c2809e85f5720a83986fad4b0a',
              name: 'masoud456',
            },
          ],
        },
      },
    };
    useSelector.mockImplementation((callback) => callback(updatedMockAppState));
    const { waitFor } = renderHook(() => useManageHWAccounts());

    await waitFor(() => {
      expect(mockDispatch).toHaveBeenCalledTimes(1);
      expect(mockDispatch).toHaveBeenCalledWith(removeHWAccounts());
    });
  });
});
