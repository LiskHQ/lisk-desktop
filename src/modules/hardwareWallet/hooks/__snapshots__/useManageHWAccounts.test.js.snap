// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useManageHWAccounts hook stores the list of accounts in the hardware wallet 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "accounts": Array [
          Object {
            "hw": Object {
              "brand": "Ledger",
              "deviceId": "20231",
              "model": "Nano S",
            },
            "metadata": Object {
              "accountIndex": 0,
              "address": "lskdgtenb76rf93bzd56cqn6ova46wfvoesbk4hnd",
              "creationTime": "2023-02-10T13:45:03.959Z",
              "isHW": true,
              "path": "",
              "pubkey": "a07e06d3d21bfcbd6e6cb158a0adc18b6ccc92db66dc8a323c7462af9c539fab",
            },
          },
        ],
        "type": "STORE_ACCOUNTS",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
